src/keyszer/lib/key_context.py

```diff
@@ -2,9 +2,21 @@
 
 
 class KeyContext:
-    def __init__(self, device_name):
-        self._device_name = device_name
+    def __init__(self, device):
         self._X_ctx = None
+        leds_list = []
+
+        # Must declare these here or app will crash if keyboard device hasn't been "grabbed" yet
+        self._capslock_state = ""
+        self._numlock_state = ""
+
+        # Check for actual device name being present before using evdev's ".leds()" method
+        # or device.name (doesn't like strings)
+        if not device == "":
+            self._device_name = device.name
+            leds_list = device.leds()
+            self._capslock_state = "ON" if 1 in leds_list else "OFF"
```

Comment:
This code should move into the property getters. It's a code smell to do a bunch of work in the initializer that we don't need to do. 

Also we should use True False here. 

Maybe naming like `is_numlock`, etc.

Reply:
OK, changing back to True/False seems to keep working (when I remembered to change the criteria in the config). 

Naming of `is_numlock` doesn't really match the other properties, to me. Would `numlock_on` be OK? Using `numlock_on is True` as a test criteria seems very intuitive for the user. 

I tried to minimize what was being done in the initializer by only running `device.leds()` and populating `leds_list` once, and of course not running any of that at all until there is a valid device. Moving things into the individual property getters seems to introduce a lot of redundancy and need to declare globals and such due to things being undefined inside the getter. If you're sure that's the better way to do things it will take me a while to figure out how to get it all back into a working state.

Reply:
> Would numlock_on be OK?

Sure, lets try that.  I like `is_` for booleans but the double is might be a little weird in python.

Reply:
Done moving pretty much everything out of `init`, and it resolved a lot. See the other review thread.